samples/transform_dialect/transform_library.mlir

```diff
@@ -0,0 +1,52 @@
+  transform.named_sequence @custom_matmul(%matmul: !transform.any_op {transform.consumed}) {  
+    %1 = transform.structured.generalize %matmul : (!transform.any_op) -> !transform.any_op
+    transform.print {name = "Setting matmul strategy to default"}
+    transform.yield
+  }
+
+  // Send it down subgroup reduce.
+  transform.named_sequence @use_subgroup_reduce(%reduce: !transform.any_op {transform.readonly}) {  
+    %variant_op = transform.get_parent_op %reduce {op_name = "hal.executable.variant"} : (!transform.any_op) -> !transform.any_op
+    %lowering_config = transform.param.constant #iree_codegen.lowering_config<tile_sizes = [[8, 0], [1, 0], [0, 0, 4]]> -> !transform.any_param
+    transform.annotate %reduce "lowering_config" = %lowering_config : !transform.any_op, !transform.any_param
+    %exports = transform.structured.match ops{["hal.executable.export"]} in %variant_op : (!transform.any_op) -> !transform.any_op
+    %subgroup_reduce = transform.param.constant #iree_codegen.translation_info<SPIRVBaseVectorize> -> !transform.any_param
+    %workgroup_size = transform.param.constant [16 : index, 1 : index, 1 : index] -> !transform.any_param
+    transform.annotate %exports "translation_info" = %subgroup_reduce : !transform.any_op, !transform.any_param
+    transform.annotate %exports "workgroup_size" = %workgroup_size : !transform.any_op, !transform.any_param
+    transform.print {name = "Setting reduce strategy to base vectorize"}
+    transform.yield
+  }
+
+  //===------------------------------------------------------===
+  // Matchers
+  //===------------------------------------------------------===
+  transform.named_sequence @match_matmul(%matmul: !transform.any_op {transform.readonly}) -> (!transform.any_op) {
+    transform.match.operation_name %matmul ["linalg.matmul"] : !transform.any_op
+    transform.yield %matmul : !transform.any_op
+  }
+
+  transform.named_sequence @match_reduce(%reduce: !transform.any_op {transform.readonly}) -> (!transform.any_op) {
+    transform.match.operation_name %reduce ["linalg.generic"] : !transform.any_op
+    %matched = transform.match.structured failures(propagate) %reduce : (!transform.any_op) -> (!transform.any_op) {
+    ^bb1(%arg1: !transform.any_op):
+      %c2 = transform.param.constant 2 : i64 -> !transform.param<i64>
+      %rank = transform.match.structured.rank %arg1 : (!transform.any_op) -> !transform.param<i64>
+      transform.match.param.cmpi eq %rank, %c2 : !transform.param<i64>
+  
+      transform.match.structured.dim %arg1[-1] {reduction} : !transform.any_op
+      transform.match.structured.yield %arg1 : !transform.any_op
+    }
+    transform.yield %matched : !transform.any_op
+  }
+
+  transform.named_sequence @transform_main(%variant_op: !transform.any_op {transform.consumed}) {  
+    transform.foreach_match in %variant_op
+        @match_matmul -> @custom_matmul,
+        @match_reduce -> @use_subgroup_reduce
+      : (!transform.any_op) -> (!transform.any_op)
```

Comment:
Unfortunately what I wanted to do doesn't seem to quite work yet. Essentially `foreach_match` walks the payload IR and let's you do a match (and rewrite), but only on that operation. If I were to try to codegen the variant then the foreach will keep walking the IR and crash, so annotation is the best I can do for now if I want more than one strategy colocated within the same spec.

Also AFAICT there isn't a `transform.if` or `transform.switch` + the matcher dialect seems a bit tuned for structured ops, so this seems like a good example case for some missing upstream transform ops; once we have those I can update some of these samples.

Reply:
Sounds good. I am just looking for the removal of setting user configs from CPU and SPIR-V backends (and use the materialize user encoding path here). After that this is good to go.

Reply:
oh I just updated that. This comment was unrelated and more of just a PSA.